Foundation/CPCache.j

```diff
@@ -0,0 +1,350 @@
+    int                     _currentPosition;
+    int                     _countLimit;
+    int                     _totalCostLimit;
+    id <CPCacheDelegate>    _delegate;
+
+    // Set of bit to determine which delegate has to respond
+    unsigned                _implementedDelegateMethods;
+}
+
+
+#pragma mark Initialization
+
+/*!
+    Initializes the cache with default values
+    @return the initialized cache
+*/
+- (id)init
+{
+    self = [super init];
+
+    if (self)
+    {
+        _items = [[CPDictionary alloc] init];
+        _currentPosition = 0;
+        _countLimit = 0;
+        _totalCostLimit = 0;
+        _delegate = nil;
+    }
+
+    return self;
+}
+
+
+#pragma mark Managing cache
+
+/*!
+    Returns the object which correspond to the given key
+    @param aKey the key for the object's entry
+    @return the object for the entry
+*/
+- (id)objectForKey:(id)aKey
+{
+    return [[_items objectForKey:aKey] object];
+}
+
+/*!
+    Adds an object with default cost into the cache.
+    @param anObject the object to add in the cache
+    @param aKey the object's key
+*/
+- (void)setObject:(id)anObject forKey:(id)aKey
+{
+    [self setObject:anObject forKey:aKey cost:0];
+}
+
+/*!
+    Adds an object with a cost into the cache.
+    @param anObject the object to add in the cache
+    @param aKey the object's key
+    @param aCost the object's cost
+*/
+- (void)setObject:(id)anObject forKey:(id)aKey cost:(int)aCost
+{
+    // Check if the key already exist
+    if ([_items objectForKey:aKey])
+        [self removeObjectForKey:aKey];
+
+    // Add object
+    [_items setObject:[_CPCacheItem cacheItemWithObject:anObject cost:aCost position:++_currentPosition] forKey:aKey];
+
+    // Clean cache to satisfy condition (< totalCostLimit & < countLimit) if necessary
+    [self _cleanCache];
+}
+
+/*!
+    Removes the object from the cache for the given key.
+    @param aKey the key of the object to be removed
+*/
+- (void)removeObjectForKey:(id)aKey
+{
+    // Call delegate method to warn that the object is going to be removed
+    if (_implementedDelegateMethods & CPCacheDelegate_cache_WillEvictObject)
+        [_delegate cache:self willEvictObject:[[_items objectForKey:aKey] object]];
+
+    [_items removeObjectForKey:aKey];
+}
+
+/*!
+    Removes all the objects from the cache.
+*/
+- (void)removeAllObjects
+{
+    _currentPosition = 0;
+
+    // Call delegate method to warn that the objects are going to be removed
+    if (_implementedDelegateMethods & CPCacheDelegate_cache_WillEvictObject)
+    {
+        var enumerator = [_items keyEnumerator],
+            value;
+
+        while (value = [enumerator nextObject])
+            [_delegate cache:self willEvictObject:[[_items objectForKey:value] object]];
+    }
+
+    [_items removeAllObjects];
+}
+
+
+#pragma mark Accessors
+
+/*!
+    Returns the count limit of the cache
+*/
+- (int)countLimit
```

Comment:
you could use for instance `@accessors(property=countLimit)` in the class ivar list instead of adding all these methods manually.

You can also declare a delegate and override the setter or getter if you need to, that way you get code for free, and you can add your own only where needed